### PR TITLE
SI-8323 Erase constant types for better double-def checks

### PR DESCRIPTION
**Pull request by retronym**
Historically, Constant Types in method signatures have not been
erased. Rather, this was done in the backend.

This state of affairs leads to wrong decisions when the erasure
phase makes decisions about whether a pair of methods will clash
in bytecode.

This commit changes `ErasureMap` to take the underlying type of a
constant type.

This change obviates the fix for SI-4753, dcc07bd9f.

I've added test to show that constants are still inlined.

I believe this will answer the question in the backend about why
existential types make it that far; they come from the underlying
of a constant type representing a class literal. I'll leave it to
@lrytz to verify that the code in question is now dead and to remove
it in his backend work.

This change should be binary compatible with 2.11.0. Even though we
changed erasure, the combination of erasure and bytecode generation
lead to the same results.

Review by @lrytz @gkossakowski

I ran this patch by @odersky last Friday. He said it looks correct
so long as we verify it doesn't affect behaviour of inlining. He
showed my some code in Dotty that actually followed the new approach.

However, when I look at dotty master, I see that `ConstantTypes`
are not erased: https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/transform/Erasure.scala#L129.

So, clarification by @odersky, too.

### CODE REVIEW

**Comment by retronym** [comment]
PLS REBUILD ALL

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 1]
```diff
@@ -115,7 +115,8 @@ trait Erasure {
 
     def apply(tp: Type): Type = tp match {
       case ConstantType(_) =>
-        tp
+         // q"classOf[String]".tpe.deconst gives Class[String], we have to recurse to get to `Class[_]`
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:118]: Why does `t"Class[String]".deconst` evaluate to t"Class[_]"? What's the representation of `t"Class[String]"` here?

Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:118]: Maybe the REPL says it better than I managed:

``` scala
scala> val tp = ConstantType(Constant(typeOf[String]))
tp: $r.intp.global.UniqueConstantType = Class[String](classOf[java.lang.String])

scala> tp.deconst
res6: $r.intp.global.Type = Class[String]

scala> erasure.scalaErasure(res6)
res7: $r.intp.global.erasure.global.Type = Class
```

Comment (lrytz) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:118]: It helps, maybe also add the information that Constant.tpe creates the `Class` type ref

``` scala
scala> Constant(typeOf[String]).tpe
res0: $r.intp.global.Type = Class[String]
```

**Comment by lrytz** [comment]
LGTM

**Comment by retronym** [comment]
I need to compensate for this change in `Mixin` which special cases mixed in members with constant type results:

``` scala
        assert(getter.isGetter)
        val readValue = getter.tpe match {
          // A field "final val f = const" in a trait generates a getter with a ConstantType.
          case MethodType(Nil, ConstantType(c)) =>
            Literal(c)
```

Closing this PR in the meantime.

**Comment by lrytz** [comment]
It might be good to look at references of `ConstantType` in all phases after erasure, there's for example [PostErasure](https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/transform/PostErasure.scala#L12)

**Comment by retronym** [comment]
Reopening so the kitteh can work on a test run while we go for a :beer: run.

**Comment by retronym** [comment]
I might retarget this against 2.12.x, btw.

**Comment by gkossakowski** [comment]
@retronym: what's the status of this PR? Are planning to retarget it against 2.12.x?

**Comment by retronym** [comment]
Yep. As soon as I squash the residual issues I'll do that.

**Comment by retronym** [comment]
BTW, despite the slight difficulty this change causes to a few phases after erasure, I still believe it is the most principled fix for the bug: erasure should erase _all the way_ to JVM signatures, and later phases should use phase time-travel to make decisions based on the "constant-cy" of symbol's infos.

**Comment by lrytz** [comment]
+1

**Pull request closed without merging**